### DISCUSSION TITLE
Help with figuring out what's wrong with Svelte grammar

### DISCUSSION THREAD

**Discussion by kindoflew** [github-linguist/linguist#6164] [Q&A]
I opened a PR in the upstream Svelte grammar to attempt to fix broken syntax highlighting (https://github.com/sveltejs/svelte-atom/pull/21), but without any solid way to test it, we were left waiting for the new release of `linguist` to see if it worked or not.

Another dev opened a PR here (https://github.com/github/linguist/pull/6144) before the new release, but closed it, because we determined that the fix in `svelte-atom` *should* have worked.

That dev made a fork of the community fork of Lightshow (https://umanghome.github.io/NovaLightshow/) that allowed us to test the grammar. Using that, I found that `NovaLightshow` was getting tripped up by `svelte-atom`'s `.cson` grammar -- it was throwing a `YAMLError`. If I converted it manually to JSON (I used this site: https://www.convertsimple.com/convert-cson-to-json/) and pasted it in the Grammar box, syntax highlighting worked.

It was also my understanding that `linguist` converts `.cson` files to `.json` when building its grammars: https://github.com/github/linguist/blob/32ec19c013a7f81ffaeead25e6e8f9668c7ed574/tools/grammars/compiler/proto.go#L65-L69 So I assumed that meant the "fixed" upstream grammar would work once converted to JSON.

However, after the release of `7.24.0`, Svelte syntax highlighting is still broken on GitHub. Is there anyone who knows more about this stuff who could possibly help figure out what is going wrong here? Thanks!

**Comment by lildude** [comment]
I've done some digging and the issue is going to be down to GitHub's old highlighting engine which still parses these TextMate-based grammars. 

The problem is it doesn't expand scope matches and instead treats the `$1` in places like [this](https://github.com/sveltejs/svelte-atom/blob/865a55f096ed0d60a316234f55f6f356abd5cd1f/grammars/svelte.cson#L606) literally:

```cson
'tags-lang':
    begin: '<(script|style|template)'
    end: '</\\1\\s*>|/>'
    beginCaptures:
      '0':
        patterns: [include: '#tags-start-node']
    endCaptures:
      '0':
        patterns: [include: '#tags-end-node']
    name: 'meta.$1.svelte'.  #<---------------- HERE
    patterns: [
      begin: '\\G(?=\\s*[^>]*?(type|lang)\\s*=\\s*([\'"]|)(?:text/)?(\\w+)\\2)'
      end: '(?=</|/>)'
      name: 'meta.lang.$3.svelte'
      patterns: [include: '#tags-lang-start-attributes']
    ,
      include: '#tags-lang-start-attributes'
    ]
```

This means it will not match any of your injections which are expecting the `$1` to be replaced with `script|style|template`. 

This is news to me. It's also unfortunately been like this for well over 8 years and unfortunately isn't going to change as this old parser is no longer getting internal funding for fixes or improvements in favour of the newer tree-sitter based highlighting.

This means the only solution is to be explicit with the scope names and not rely on the regex matches, or wait for the Svelte tree-sitter grammar (I think [this](https://github.com/Himujjal/tree-sitter-svelte) is it) to be pulled in by the team that maintains the highlighting engines.

If the tree-sitter grammar is at least as functionally complete as the TextMate grammar (lemme know, as I don't know), I can have a word with the team and put it on their radar for consideration.

**Reply by nasso** [reply]
this is what we get for parsing html with regex...

**Reply by lildude** [reply]
I've done some quick testing in dev using the tree-sitter grammar and it looks promising and much better than the Textmate grammar:

![CleanShot 2022-11-22 at 09 45 37](https://user-images.githubusercontent.com/627280/203281409-8d8a2cd2-8dd5-4e32-b6ee-3771953e69b8.png)

I'm going to submit a PR proposing to add the Svelte tree-sitter grammar. We'll see how it goes.

**Reply by kindoflew** [reply]
oh wow, awesome! thanks so much!

**Reply by lildude** [reply]
So it turns out my local dev env was not actually pulling in the treesitter grammar when testing. It was actually pulling in the TextMate grammar from Linguist v7.20.0 which means the TextMate grammar was working properly then and started failing after that.

Really (promise this time) testing the tree-sitter grammar from https://github.com/Himujjal/tree-sitter-svelte showed it is questionably "better" (ie it's not much) than what we have now:

![CleanShot 2022-11-22 at 15 11 54](https://user-images.githubusercontent.com/627280/203350108-c6a0d7e8-76b9-46d4-b05e-0709923cc4d5.png)

So the tree-sitter grammar is definitely _not_ ready for prime time, so I will not be continuing down that avenue.

Back to the TextMate grammar...

The first time we see a change in the Svelte grammar after Linguist v7.20 was the change in https://github.com/github/linguist/pull/5963 which changed the source of the grammar. This first shipped in Linguist v7.22.0 and is likely when the breakage occurred.

Looking at the two grammars, I can see they're quite different.

It does mean we have a quick fix: revert https://github.com/github/linguist/pull/5963 and go back to the four-year old grammar from https://github.com/umanghome/svelte-atom whilst we wait for a prime-time tree-sitter grammar or for the https://github.com/sveltejs/svelte-atom grammar to update to take into account the limitations of the ancient highlighter.

Given how impactful this is, I'm leaning towards reverting back to the old grammar.

**Reply by lildude** [reply]
It's interesting to note that https://github.com/umanghome/svelte-atom was forked from https://github.com/sveltejs/svelte-atom 

Looking through the history of differences, this would be where things broke: https://github.com/sveltejs/svelte-atom/commit/d01994a8cc2a0af5867484a4d6185c81217f79ed This commit suggests this was with the adoption of the VSCode grammar and can see things switch from explicit scopes to using expanded regex matched names.

**Reply by lildude** [reply]
> Given how impactful this is, I'm leaning towards reverting back to the old grammar.

I've just been reminded that we're close to the feature freeze date for the next GHES release and I'd rather not ship with known broken syntax highlighting so I'm going to go for this option and make a patch release early next week.

If the https://github.com/sveltejs/svelte-atom grammar has been updated to address the limitations, I'll use that instead.

**Reply by lildude** [reply]
https://github.com/github/linguist/pull/6184

**Reply by kindoflew** [reply]
Sorry for the delay, Thanksgiving holiday in the US. 

I see that you've already merged the old grammar, but I agree -- there's a few bugs in it still, but it's better than the completely broken highlighting we had.

I haven't gotten to open a PR in `svelte-atom` yet, but since we have syntax highlighting *mostly* working again, it doesn't seem like a time sensitive issue anymore. I'll either open one there eventually or reach out to the `svelte` team to see if they have any other ideas (like eventually maintaining an official TreeSitter grammar under the `sveltejs` org). Either way, probably expect a PR from me sometime before the next, next GHES release I guess 😅.

Thanks so much again for getting involved in this!
